semgrep/semgrep/semgrep_main.py

```diff
@@ -244,52 +345,47 @@ def main(
     except FilesNotFoundError as e:
         raise SemgrepError(e)
 
-    join_rules, rest_of_the_rules = partition(
-        lambda rule: rule.mode == JOIN_MODE,
-        filtered_rules,
-    )
-    dependency_aware_rules = [
-        r for r in rest_of_the_rules if r.project_depends_on is not None
-    ]
-    filtered_rules = rest_of_the_rules
+    # Initialize baseline before running rules to fail early on bad args
+    baseline_handler = None
+    if baseline_commit:
+        try:
+            baseline_handler = BaselineHandler(baseline_commit)
+        # TODO better handling
+        except Exception as e:
+            raise SemgrepError(e)
 
     core_start_time = time.time()
-    # actually invoke semgrep
-    (rule_matches_by_rule, semgrep_errors, all_targets, profiling_data,) = CoreRunner(
+    core_runner = CoreRunner(
         jobs=jobs,
         timeout=timeout,
         max_memory=max_memory,
         timeout_threshold=timeout_threshold,
         optimizations=optimizations,
-    ).invoke_semgrep(target_manager, profiler, filtered_rules)
-
-    if join_rules:
-        import semgrep.join_rule as join_rule
-
-        for rule in join_rules:
-            join_rule_matches, join_rule_errors = join_rule.run_join_rule(
-                rule.raw, [Path(t) for t in target_manager.targets]
-            )
-            join_rule_matches_by_rule = {Rule.from_json(rule.raw): join_rule_matches}
-            rule_matches_by_rule.update(join_rule_matches_by_rule)
-            output_handler.handle_semgrep_errors(join_rule_errors)
+    )
 
-    if len(dependency_aware_rules) > 0:
-        import semgrep.dependency_aware_rule as dep_aware_rule
+    (
+        rule_matches_by_rule,
+        semgrep_errors,
+        all_targets,
+        profiling_data,
+    ) = run_rules(filtered_rules, target_manager, core_runner, output_handler)
+    profiler.save("core_time", core_start_time)
+    output_handler.handle_semgrep_errors(semgrep_errors)
 
-        for rule in dependency_aware_rules:
+    # Run baseline if needed
+    if baseline_handler:
+        logger.info(f"Running baseline scan with base set to: {baseline_commit}")
+        with baseline_handler.baseline_context():
```

Comment:
Does this need to be in a try-catch? Or are errors here handled somewhere else? If so please update main docstring

Reply:
Good call